soroban-env-host/src/host/prng.rs

```diff
@@ -98,8 +103,64 @@ impl Prng {
+    // method says:
+    //
+    //     PRNG implementations are allowed to assume that bits in the seed
+    //     are well distributed. That means usually that the number of one and
+    //     zero bits are roughly equal, and values like 0, 1 and (size - 1)
+    //     are unlikely. Note that many non-cryptographic PRNGs will show poor
+    //     quality output if this is not adhered to. If you wish to seed from
+    //     simple numbers, use seed_from_u64 instead.
+    //
+    // ChaCha _is_ a cryptographic PRNG, and it's hard to find any specific
+    // _requirements_ it puts on its keys, but the Salsa20 [secuity-claims
+    // document](https://cr.yp.to/snuffle/security.pdf) does have the following
+    // hints (note ChaCha is a variant of Salsa):
+    //
+    //     If the Salsa20 key k is a uniform random sequence of bytes, and the
+    //     same nonce is never used for two different messages, then the Salsa20
+    //     encryption function is conjectured to produce ciphertexts that are
+    //     indistinguishable from perfect ciphertexts, i.e., uniform random
+    //     strings independent of the plaintexts
+    //
+    // I.e. the _premise_ of its security claims are "uniform random key". Also:
+    //
+    //     Related-key attacks:
+    //
+    //     The standard solutions to all the standard cryptographic problems --
+    //     encryption, authentication, etc. -- are protocols that do not allow
+    //     related-key attacks on the underlying primitives. I see no evidence
+    //     that we can save time by violating this condition. The reader might
+    //     guess that Salsa20 is highly resistant to related-key attacks; but I
+    //     simply don’t care.
+    //
+    // I believe "biased key" attacks are a subset of "related key" attacks and
+    // so we should at least assume that ChaCha is designed with the assumption
+    // that the key is not biased.
+    //
+    // It's also worth noting that the `SeedableRng::seed_from_u64` method seeds
+    // a different PRNG (PCG) with the u64, and then uses its output as the key
+    // for ChaCha, and this appears to be a fairly common approach in other
+    // libraries I've surveyed. I'm not enough of a cryptographer to understand
+    // exactly why this helps, considering ChaCha is itself a hash construct
+    // that's supposed to diffuse all entropy in every input (key & nonce) bit
+    // to every output bit, but it seems to be a common pattern.
+    //
+    // Given all this, we're going to run any user-provided seed through SHA256
+    // before using it as a ChaCha key. In production we should always be
+    // receiving a seed that's an output from SHA256 of ledger and transaction
+    // context information, but this adds a layer of defense against the
+    // possibility that we're given a nonuniform seed (which happens in testing
+    // at very least, and users might also manually reseed with a biased seed).
```

Comment:
I would not use SHA-256 for this purpose.

The "best possible" property you could ask for here is: given an input with `k` bits of entropy, you want to output a 256-bit string with (negligibly less than) `min(256, k)` bits of entropy, with the additional property that an adversary has (negligibly more than) `2^-k` chance of distinguishing the output from uniform. In other words: you want to conserve the input's entropy, but get an output that's as close as possible to uniform.

This isn't what SHA-2 was designed to do, and to my knowledge there's no formal analysis of SHA-2 that would give this property other than simply by regarding SHA-2 as a random oracle. In other words: the only way to show that SHA-2 is good enough for this task is simply to assume that it is.

But the good news is that there has been a lot of work on ways to implement randomness extractors using functions like SHA-2---and it turns out to be pretty straightforward! If you only read one paper on this topic, read Krawczyk's [analysis of HKDF](https://eprint.iacr.org/2010/264.pdf). It's probably also worthwhile to read Hugo's notes about HKDF in [RFC5869](https://www.rfc-editor.org/rfc/rfc5869).

---

The fix, as you may have guessed, is to use something like HKDF. In fact, you could just use RFC5869 directly:

1. Sample a fixed, public 32-byte random value, `salt`. Hard-code this value in the source code.
2. Fix some domain separator string, `info`. Perhaps something like `soroban::new_prng_from_seed` would work.
3. Instead of computing `sha256(input)`, compute `hkdf-sha256(salt, info, input, 32)`.

The above requires 4 SHA-256 invocations (a.k.a. two HMAC-SHA-256 invocations). It is almost certainly safe to cut this in half:

1. Sample `salt` as above.
2. Instead of computing `sha256(input)`, compute `hmac_sha256(key=salt, message=input)`.

There's nothing cryptographically dangerous about this, but it does mean that you lose the degree of freedom that the `info` string gives you. So if you, say, wanted to derive two independent keys from the same user input, that's slightly harder / less efficient. But if you don't think that's a downside in this context, HMAC is good enough.

---

There may be some political question around selecting the random seed: extremely paranoid users might suspect that you've somehow cooked the seed to your advantage (never mind that this is unlikely to be possible).

So, if necessary, you could select the seed in a "nothing up my sleeve" way by computing, e.g., `hmac_sha256(key=0, message=<first million digits of pi as ascii>)`.

Reply:
@kwantam Ok, I think I will just go with the simpler/cheaper "just an hmac with a fixed key" option; is it acceptable to make the fixed key itself the output of a process a bit simpler than generating a million digits of pi? i.e. can I set it to `sha256("nothing up our sleeve")` or so (assuming I check to see that that has a fairly good balance of zeroes and ones)? Is there some standard sense of what makes a "public random value" appropriate in this role?

Reply:
(Actually I just tried and it's fairly trivial to write a python program that calculates hash-of-pi-digits so if that is in any way more robust to assumptions I'm happy to use that too; whatever will raise fewer eyebrows)

Reply:
This is more a political than a technical decision---I'm sure that `sha256("nothing up our sleeve")` would be perfectly fine to use as a salt.

If someone is seriously intent on sniffing out a nonexistent conspiracy, any choice you make here will be insufficient. But it seems like "we chose a large number of digits from a well-known constant" is pretty hard to find evil intent in. (I guess there's still the "ah but why a million and not a thousand digits" style of gotcha...)

Reply:
Ah, sorry, to answer your first question: there's nothing in particular that makes the choice of one salt better than another, modulo the nothing-up-my-sleeve question and assuming that the choice is reasonably close to uniform.

In an ideal world you'd choose a value that other folks aren't likely to choose (because this makes your KDF independent of their KDF). Who knows whether HMAC of the first million digits of pi with the zero key meets that criterion. Maybe HMAC on "first million digits of pi" as the data with "hash of the [millionth stellar ledger](https://horizon.stellar.org/ledgers/1000000)" as the key does a better job on the uniqueness axis. But now again you've got to deal with people deciding there really is something up your sleeve...

Reply:
Thanks @kwantam for all the help! I've updated to a fixed-salt-HMAC. Could you give one last glance at https://github.com/stellar/rs-soroban-env/pull/1232/commits/d8bc65dc91c6d77d57d747b96462cf434898bf46#diff-cc7d8d1f0039fd49ba0a0275f27a4e866fb7e9c800519d2e02020b79f1045c16R244-R287 to double check that I'm not saying or doing something egregiously wrong?